### DISCUSSION TITLE
Distinguish types of messages

### DISCUSSION THREAD

**Discussion by Rim-777** [visciang/telegram#170] [General]
Hello, I'm curious if there are any docs where I can find how to distinguish message types e.g. voice, file, audio, etc. 
for  example, In Python libs, it can be done by putting a specific type in the message handler., like  this:  
  ```
    app.add_handler(MessageHandler(filters.TEXT, handle_message))
    app.add_handler(MessageHandler(filters.Document.ALL | filters.PHOTO, handle_file))
    app.add_handler(MessageHandler(filters.VOICE, handle_voice))
```

Are there some specific types like that for handle_update?  

Thank you very much in advance

**Comment by visciang** [comment]
Hi @Rim-777,

the overall philosophy of the API is "a light layer over the Telegram Bot API HTTP-based interface, it does not expose any  (data)binding".

So there isn't a custom abstraction over the data exposed in the Telegram JSON API.
Any filtering can be simply "data driven", very elixir-idiomatic - what you see is what you get, no need to abstract or "obfuscate" anything over the Telegram API JSON object.

For example, let's see the [`getUpdates`](https://github.com/visciang/telegram?tab=readme-ov-file#getupdates) request.

The method is [documented](https://core.telegram.org/bots/api#getupdates) to return a list of [`Update`](https://core.telegram.org/bots/api#update) objects.

It is essentially a choice of `message` | `edited_message` |  `channel_post` ... [`Messages`](https://core.telegram.org/bots/api#message).

And a `Message` can be a `text String` | [`audio Audio`](https://core.telegram.org/bots/api#audio) | [`document Document`](https://core.telegram.org/bots/api#document) | [`video Video`](https://core.telegram.org/bots/api#video) | [`voice Voice`](https://core.telegram.org/bots/api#voice) etc etc.

So just apply erlang pattern match or basic Access | Enum filters, to dispatch to specific handlers.
There are some examples under https://github.com/visciang/telegram/tree/master/example.
